### PR TITLE
Nullkiller ai

### PR DESCRIPTION
**Pull request by nullkiller**
Finally I am ready to merge Nullkiller AI to develop and continue it there.
AI switcher works.
There are some changes in VCMI_lib which require some attention from reviewers.
Added Intel TBB to ease async tasks processing.

### CODE REVIEW

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -57,9 +57,7 @@ set(test_SRCS
  		spells/targetConditions/SpellEffectConditionTest.cpp
  		spells/targetConditions/TargetConditionItemFixture.cpp
 		
-		vcai/mock_ResourceManager.cpp
```
Comment (nullkiller) [inline test/CMakeLists.txt:60]: Need to revert these changes.

File: lib/mapObjects/CommonConstructors.h [hunk 2]
```diff
@@ -171,6 +171,8 @@ class DLL_LINKAGE CBankInfo : public IObjectInfo
 	CBankInfo(const JsonVector & Config);
 
 	TPossibleGuards getPossibleGuards() const;
+	TResources getPossibleResourcesReward() const;
+	std::vector<CStackBasicDescriptor> getPossibleCreaturesReward() const;
```
Comment (nullkiller) [inline lib/mapObjects/CommonConstructors.h:175]: This is an additional API for AI to understand what object can give when visited.

File: lib/mapObjects/CArmedInstance.cpp [hunk 3]
```diff
@@ -89,19 +89,24 @@ void CArmedInstance::updateMoraleBonusFromArmy()
 			factionsInArmy -= mixableFactions - 1;
 	}
 
+	std::string description;
```
Comment (nullkiller) [inline lib/mapObjects/CArmedInstance.cpp:92]: Avoid changing shared variable. b->description directly. Not sure why boost::algorithm::trim(b->description) crashes in multithreading environment but it does.

File: lib/CCreatureHandler.cpp [hunk 4]
```diff
@@ -181,6 +181,18 @@ void CCreature::fillWarMachine()
 	warMachine = ArtifactID::NONE; //this creature is not artifact
 }
 
+bool CCreature::DisableChildLinkage = false;
+
+void CCreature::newChildAttached(CBonusSystemNode *child)
+{
+	if(!DisableChildLinkage) CBonusSystemNode::newChildAttached(child);
```
Comment (nullkiller) [inline lib/CCreatureHandler.cpp:188]: This is hack. I am not sure how to do it in a bit more accurate way. The idea is that we need to somehow disable linking temporary bonus nodes from their non-temporary parent nodes. For instance TemporaryArmy has stacks which are also temporary but they are referenced from CCreature which belongs to global three. This backward link looks useless for temporary bonus nodes which leave just a short period of time especially taking into account that CCreature can not (at least should not) change during that period.

Comment (nullkiller) [reply lib/CCreatureHandler.cpp:188]: Probably I can try to introduce some virtual bool isTemporaryNode() to IBonusBearer instead.

File: client/Graphics.cpp [hunk 5]
```diff
@@ -121,17 +122,15 @@ void Graphics::initializeBattleGraphics()
 }
 Graphics::Graphics()
 {
-	#if 0
+	#if 1
```
Comment (alexvins) [inline client/Graphics.cpp:125]: This was disabled due to issues with SDL, please revert for now

File: CI/mxe/before_install.sh [hunk 6]
```diff
@@ -3,34 +3,36 @@
...
-mxe-$MXE_TARGET-sdl2-gfx \
-mxe-$MXE_TARGET-sdl2-image \
-mxe-$MXE_TARGET-sdl2-mixer \
-mxe-$MXE_TARGET-sdl2-ttf \
-mxe-$MXE_TARGET-ffmpeg \
-mxe-$MXE_TARGET-qt \
-mxe-$MXE_TARGET-qtbase
+if true; then
+	# MXE repository was too slow for Travis far too often
+	wget https://github.com/nullkiller/vcmi-deps-mxe/releases/download/2021-01-22/mxe-$MXE_TARGET-2021-01-22.tar
```
Comment (alexvins) [inline CI/mxe/before_install.sh:8]: This should be pushed to main repo

**Review by alexvins — CHANGES_REQUESTED** [review]

File: lib/mapObjects/CommonConstructors.cpp [hunk 7]
```diff
@@ -412,34 +412,68 @@ TPossibleGuards CBankInfo::getPossibleGuards() const
...
+	return TResources();
+}
+
+std::vector<CStackBasicDescriptor> CBankInfo::getPossibleCreaturesReward() const
+{
+	std::vector<CStackBasicDescriptor> aproximateReward;
+
+	for(const JsonNode & configEntry : config)
+	{
+		const JsonNode & guardsInfo = configEntry["creatures"];
```
[the remark this answers is not in the document]
Comment (nullkiller) [reply lib/mapObjects/CommonConstructors.cpp:434]: Will check it

File: CI/mxe/before_install.sh (hunk 6 above)
Comment (nullkiller) [reply CI/mxe/before_install.sh:8]: It needs to be released. And I have no access to the official repo to make a release or at least push a tag so that I can try github-release utility. You can check that in my repo my commit is tagged and a release is created. Was not able to push a tag using a token provided in vcmi-deps-windows.

File: AI/VCAI/VCAI.cpp [hunk 8]
```diff
@@ -1,2860 +1,2860 @@
-/*
```
[the remark this answers is not in the document]
Comment (nullkiller) [reply AI/VCAI/VCAI.cpp:1]: Looks like yes. I pushed it with core.autocrlf=true and locally I have it with crlf

File: AI/Nullkiller/AIGateway.cpp [hunk 9]
```diff
@@ -0,0 +1,1668 @@
...
+	boost::unique_lock<boost::mutex> lock(mx);
+	assert(vstd::contains(remainingQueries, queryID));
+	std::string description = remainingQueries[queryID];
+	logAi->debug("Attempted answering query %d - %s. Request id=%d. Waiting for results...", queryID, description, answerRequestID);
+	requestToQueryID[answerRequestID] = queryID;
+}
+
+void AIStatus::receivedAnswerConfirmation(int answerRequestID, int result)
+{
+	assert(vstd::contains(requestToQueryID, answerRequestID));
```
[the remark this answers is not in the document]
Comment (nullkiller) [reply AI/Nullkiller/AIGateway.cpp:1617]: It is mostly copy/paste from VCAI.cpp. I will add lock here Should I add one for original VCAI.cpp also.

File: lib/mapObjects/CommonConstructors.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (nullkiller) [reply lib/mapObjects/CommonConstructors.cpp]: You were right. Fixed.

File: lib/CCreatureHandler.cpp (hunk 4 above)
Comment (nullkiller) [reply lib/CCreatureHandler.cpp:188]: It does not look convenient to use something like StackWithBonuses. It acts as a proxy while I want to create brand new CArmedInstance to leverage from existing code for morale calculation. And this instance might consist of stacks which does not exist at all like creatures available to buy.

Comment (nullkiller) [reply lib/CCreatureHandler.cpp]: Reworked it to temporary bonus node concept

File: CI/mxe/before_install.sh (hunk 6 above)
Comment (nullkiller) [reply CI/mxe/before_install.sh]: Published releases for mxe and vcpkg

**Comment by nullkiller** [comment]
The history was rewritten and now is rebased on top of latest develop. Some additional testing is required before it can be merged.

**Comment by nullkiller** [comment]
@alexvins, do you have any idea when this PR can be merged?
